### ISSUE TITLE
Allow users to manually re-order tabs

### ISSUE THREAD

**Issue by vesta0** [mozilla-mobile/fenix#10344]
## User Story
As a user with a lot of open tabs, I want to be able to easily locate the last tab I was on and re-order tabs in a way that fits my needs, so I don't have to spend time looking for a specific tab. 

### Acceptance Criteria
* I can easily find and open the last tab I was on
* I can easily re-order tabs to where I want them to be

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-411)

**Comment by ekager** [comment]
Relevant: https://github.com/mozilla-mobile/android-components/issues/5178
We are going to be working on this very soon as part of Sync Tab work

**Comment by liuche** [comment]
Currently we put the most recent tabs on bottom, highlight the current tab, and will scroll to tab.

**Comment by yoasif** [comment]
> Currently we put the most recent tabs on bottom, highlight the current tab, and will scroll to tab.

This matches desktop, but I use https://addons.mozilla.org/firefox/addon/always-right/ (`browser.tabs.insertAfterCurrent:true`) to get tabs adjacent to the position I am at when opening a new tab. 

Fenix already opens tabs in an adjacent position when opening a new tab from an existing tab (spawning a tab). I think it would be nice to have the behavior of `browser.tabs.insertAfterCurrent:true` in Fenix by default. 

It also makes more sense with gestures - imagine if a user swipes left on a tab and doesn't find the tab that they just came from (by opening a new tab). 

Seems like that is also filed here: https://github.com/mozilla-mobile/fenix/issues/10365

**Comment by liuche** [comment]
We've done this for MVP, but we'll need to talk about how to handle the different paradigms of: self-ordered tabs vs recent tabs

**Comment by tallus** [comment]
https://github.com/mozilla-mobile/fenix/issues/3125 is also related. As a disabled user with working memory issues I rely a lot on visual/spatial memory to compensate. The current behaviour of opening links next to the parent breaks this as the place/ordering in no longer fixed and order and chronology are no longer related. For this reason I have always disabled the current behaviour in fennix when it has been present in other browsers.

Can the b:a11y tag be added (for cognitive disabilities).

**Comment by tomchiverton** [comment]
There's also no scroll bar on the column list of tabs, so as soon as full, new tabs invisible

**Comment by liuche** [comment]
@vesta0 will break this out into the actionable stages and then add them back into the sprint.

**Comment by liuche** [comment]
Just checking whether grid view is in November or October release - I see it in November in the release tracking sheet.

**Comment by liuche** [comment]
Also see #12243 with question around tabs opened by "open link in new tab"

**Comment by liuche** [comment]
We discussed what this would entail, and there's some significant AC work because the tabs tray is mostly in AC. We would also need to add a concept of tab order (which currently does not exist).

**Comment by liuche** [comment]
@csadilek we'll discuss this in backlog grooming on thursday, and we should talk about what the stages/complexity are for this issue so we could get a good size estimate. Is likely something that we'll want to pick up next sprint.

**Comment by Poopooracoocoo** [comment]
In the other issue we were discussing when drag handles could show. We liked the idea of them showing up in selection mode but selection mode isn't necessary in its current state as you can only add those tabs to a collection. You can still access the selection mode by going to save to collection in the menu. 

Therefore, removing access to the selection mode by tapping and holding in favour of dragging and dropping sounds good, unless there's plans to add more functionality to the selection mode.

I'll also reference mozilla-mobile/android-components#8376 as I did with the issue that this supersedes.

**Comment by madb1lly** [comment]
Hi all,
I'm missing this more than ever now I'm trying out grid view.

@Poopooracoocoo, the selection method could also be used for bulk bookmarking, bulk send-to-device, bulk share to other app, etc. and I presume that's why it exists, not only for save to collection.

Cheers 🙂

**Comment by Poopooracoocoo** [comment]
>the selection method could also be used for bulk bookmarking, bulk send-to-device, bulk share to other app, etc. and I presume that's why it exists, not only for save to collection.

I hope so too! I really like bulk bookmarking 🤤 

Google's struggled with this problem too with reordering bookmarks.

**Comment by liuche** [comment]
Hey @topotropic could you add mocks for this issue to the top comment? We will need the part for moving tabs around.

**Comment by ekager** [comment]
I filed a spike to hold time to discuss options for this feature (and will likely need an RFC for the AC changes) https://github.com/mozilla-mobile/fenix/issues/15928

**Comment by AdrianJPx** [comment]
Hey liuche, the trail has gone cold on this issue the past couple of months, any news on when we're likely to see the ability to manually reorder tabs added back into the app?  It's been great to get grid view back, but losing this 'skill' which the app had previously implemented so well has been so frustrating for those of us who like to keep a number of 'working tabs' open and move them around as we see fit.  Any insights into how on earth this plus grid view got removed - this was such a backward step and we expect more from Mozilla.  Every desktop browser for years now has allowed you to reorder you open tabs, I remember what such a feature first appeared.  It's such a common part of the web browsing experience I just cannot understand how a company like Mozilla would casually discard it.  Moan over, I'm sorry but it just beggars belief that they thought this was OK.  Thanks!

**Comment by issue-dispenser** [comment]
@jonalmeida Has there been any progress on this? This feature is the only issue keeping me from upgrading from Firefox 68.11.0 to Fenix.

**Comment by sagenoob** [comment]
Hello. I am a general user of firefox and not a developer or programmer. I made an account here because a moderator commented on my recent post on the support forum for firefox on android that I was welcome to comment here. 

I am EXTREMELY UPSET by not being able to reorder my tabs on android anymore. I have a chronic pain disability and mental health issues that make cognitive processes difficult for me. Not being able to reorder my tabs has been so upsetting to me. I don't understand why this was taken away, but I am advocating for this to be fixed soon. As soon as possible (hopefully that doesn't mean years from now). Not having this option has made my experience with using firefox on android miserable. 

I also wish there could be a tabgroups feature as well but I am assuming that's never going to happen. If I could just reorder the tabs that would at least help me tremendously to keep my thoughts and usage of firefox clear and allow me the freedom to focus on what I am trying to accomplish rather than being constantly confused when I am going through my tabs. 

Thank you for your consideration in this matter.

**Comment by tecknojock** [comment]
Are we ever going to see any movement on this? If this never changes, when I can no longer use Fennec, I'm going to just go back to chrome I suppose.

**Comment by AaronM04** [comment]
Some development progress on this would be greatly appreciated. Like the other commenters above, having this feature being taken away upon updating Firefox was very upsetting for me.

**Comment by mrBega** [comment]
Hello and sorry if this doesn't belong here. I'm not well versed in the ways of this forum. 

I have the problem that when I use the "undo close tab" pop up button, the tab reopenes at the top in the first position. I would like it at the bottom because that is usually the tab I close by mistake. Or even better in the spot where it was when I closed it.

Does anyone know where I should post this request? I find this forum hard to navigate on my phone.

I use the grid tab pattern NOT the list one.

And to finish of I would like to extend my deepest gratitude to everyone working on keeping the internet kinda free 💓

**Comment by gonhidi** [comment]
> I have the problem that when I use the "undo close tab" pop up button, the tab reopenes at the top in the first position. […]
> 
> Does anyone know where I should post this request?

That behaviour — when using list view — has been noted under issue #20638, so you can start by following that report. If you think it is necessary, you could comment there your experience that it also happens when viewing tabs as a grid.

(Given how new the report I mentioned is and how long Firefox has been working that way, there might be older issues for the same behaviour with a richer history; but if so, they should eventually get linked to one another.)

**Comment by hokonch** [comment]
> In planning, we decided that we need to prioritize the SessionManager/Store work over the tab reordering, https://github.com/mozilla-mobile/fenix/issues/15928#issuecomment-714699300

> We decided not to go ahead until the SessionManager -> BrowserStore refactoring is complete, as otherwise it would cause a lot of duplicate work.  https://github.com/mozilla-mobile/android-components/pull/8838#issuecomment-718941753

> After 2+ years slowly and incrementally working towards this goal, we completed the migration from the browser-session component to the browser-state component for state handling. Finally, we were able to delete browser-session and all state is now maintained and updated by the redux-like BrowserStore.  [Mozilla Android Components Blog (2021-07-05)](https://mozac.org/2021/07/05/whats-next.html)

**Comment by sagenoob** [comment]
Does that mean tab reordering will be worked on soon?

**Comment by doodhout** [comment]
Not trying to celebrate too early, but it seems that the relevant PR (https://github.com/mozilla-mobile/android-components/pull/10936) has been merged to master and is currently expected to be released in version 94.

Huzzahs are in order! Huzzah! Great job guys. 👏👏👏

**Comment by cadeyrn** [comment]
@doodhout That's at least the API part. It still needs an integration on the Fenix side. There is #21157 for that but it's still in progress, awaits a review (and depending on the results probably further iterations, especially since there are still open questions). The next version bump will be soon so I am not sure if shipping in version 94 is realistic.

**Comment by eUgEntOptIc44** [comment]
Happy halloween everyone! It's more than a month later. Any news on the status of #21157 by @ssk97 or @cadeyrn? Would highly appreciate a short update on this.

**Comment by cadeyrn** [comment]
> Any news on the status of #21157 by @ssk97 or @cadeyrn? 

I can't provide a status update, I am not involved in the development.

**Comment by ssk97** [comment]
> Happy halloween everyone! It's more than a month later. Any news on the status of #21157 by @ssk97 or @cadeyrn? Would highly appreciate a short update on this.

AFAIK they're still internally discussing how they want it to work, particularly its interactions with the tab grouping feature. That area of the code base is also changing fairly rapidly, so I think they don't want an outside contributor's work to be added while that's going on.

**Comment by saberblaze** [comment]
I have been following this issue for some time, I didn't realize that the Nightly version finally added this feature. This can be a workaround until the feature finally lands in the stable version in the future.

**Comment by issue-dispenser** [comment]
Yes, it's working for me outside of private mode (due to private mode not supporting multi-tab select?). Hopefully it's usable in private mode eventually.

**Comment by YesterKo** [comment]
How were you able to use it in Nightly? I tried to use it but wasn't able to reorder them.

**Comment by doodhout** [comment]
@YesterKo 

1. Open tab overview
2. Open Tab Settings
3. Disable Tab Grouping
4. Return to tab overview and start reordering tabs

**Comment by YesterKo** [comment]
Thank you!

**Comment by MlleBaeten** [comment]
(Im)patiently awaiting the general release of this feature! 
Not Firefox's fault specifically, since this is a general UX issue I experience across platforms, but as more and more activities move to webpages instead of applications, the fact that the action to switch between recent activities (alt+tab on windows, recent button in Android) doesn't include switching between browser tabs is a frequent annoyance. I'm adapting to the browser bar swipe for switching fenix tabs, but that's only efficient if I can move relevant tabs to be adjacent.
Long story short - we need to be able to reorder tabs to navigate efficiently! Thanks for all your work, I'm still so grateful that Mozilla exists and has entered back into relevance.

**Comment by AaronM04** [comment]
Is there an ETA on when this will land in the stable Firefox for Android?

**Comment by Dakkaron** [comment]
I'd second a wish for an ETA. From what I can see, it's been almost a year since the PR.

**Comment by lbmaian** [comment]
Might have to wait a while. It's pretty clear with the excruciatingly slow pace of updates and addon support compared to desktop Firefox, that Fenix is relatively underfunded (or perhaps funded according to its market share)

**Comment by Anutrix** [comment]
Seems it was added in https://github.com/mozilla-mobile/fenix/commit/6637f610884bf4c6826886b4a413a23f3400622e.
It looks like it's v104 beta but current version is v103.2. So little more waiting to go.
A comment from a dev/maintainer confirming this would be appreciated.

**Comment by madb1lly** [comment]
It works! 🙌

**Comment by 01tot10** [comment]
Hmm, updated Firefox motivated by @madb1lly's comment, but for my region the feature is not introduced yet. Running 103.2.0 (Build #2015895963).

**Comment by tukusejssirs** [comment]
> Seems it was added in https://github.com/mozilla-mobile/fenix/commit/6637f610884bf4c6826886b4a413a23f3400622e.
> It looks like it was added since v97 beta but current version is v103.2.
> It works in Firefox Nightly but not in Firefox yet. Not sure when it will land in the main version.
> A comment from a dev/maintainer confirming this would be appreciated.
> 

I can’t confirm this working. I use Firefox Nightly 105.0a1 (Build #2015898027), but I can’t reorder the tabs. There is no `Sort open tabs` option.

![Screenshot_20220815-084925_Firefox Nightly.jpg](https://user-images.githubusercontent.com/5111894/184589315-7a81cafe-a00f-43b0-89d3-3b9c7601e39e.jpg)

**Comment by Dakkaron** [comment]
Doesn't work for me either, though it did before... Looks like they broke that feature ...

**Comment by Dakkaron** [comment]
Doesn't work for me either, though it did before... Looks like they broke that feature ...

This commit here broke the feature: https://github.com/mozilla-mobile/fenix/commit/db0162465b42b10fbbbda2d0795b1aa0e6bcba86

@csadilek @ssk97 @pocmo

**Comment by Dakkaron** [comment]
I guess, the reason why it wasn't in regular Firefox yet was because it conflicted with tab groups and was kinda non-obvious how to turn it on. But now that tab groups appear to have been dropped, there should be nothing against enabling the tab reordering feature on regular Firefox

**Comment by Anutrix** [comment]
It's working on Grid View tabs reordering on fresh Firefox Nightly install. Haven't checked List View.

**Comment by ssk97** [comment]
> This commit here broke the feature: [db01624](https://github.com/mozilla-mobile/fenix/commit/db0162465b42b10fbbbda2d0795b1aa0e6bcba86)

Works fine for me still. What are the exact symptoms you're experiencing?

The feature flag `tabReorderingFeature` is still set to `isNightlyOrDebug` so it not working in beta or release is expected. Not sure if/when they're planning on changing that, but it's as simple as changing the variable to "true" whenever they want it in.

I suppose the comment there about it working if "tab groups are disabled" can be removed now that tab groups are gone.

**Comment by Dakkaron** [comment]
Strange. I reset cache and data, and now it's working fine. Sorry for the wrong alarm.

Since that feature has never made it to beta/release, that shouldn't cause issues for anyone.

Yeah, the wrong feature flag comment doesn't count as broken ;)

Do you know who makes the decision when these features move to the next stage?

**Comment by DreVla** [comment]
This feature is enabled in nightly already. We have a product decision to enable it in beta and release also.

**Comment by Dakkaron** [comment]
Nice! That is really good news!

Now we only need keyboard shortcuts and tablet-ui-tabs and FF for Android is back in business, regarding the biggest missing features. (At least, the biggest ones for me)

**Comment by sagenoob** [comment]
> This feature is enabled in nightly already. We have a product decision to enable it in beta and release also.

Does this mean tab reordering is coming back to the main andoird firefox app? I left a comment here months ago explaining my frustration and that I am not super tech savvy so I am unfamiliar with the lingo. 

Any guarantee and timeline? 

Thank you.

**Comment by Dakkaron** [comment]
> > This feature is enabled in nightly already. We have a product decision to enable it in beta and release also.
> 
> Does this mean tab reordering is coming back to the main andoird firefox app? I left a comment here months ago explaining my frustration and that I am not super tech savvy so I am unfamiliar with the lingo.
> 
> Any guarantee and timeline?
> 
> Thank you.

Disclaimer: Not a Mozilla employee here.

They just pushed a change that removes the feature toggle which, up until now, prevented the feature from showing up in the beta and release (which is the "real") versions. Since this is removed, the feature should show up in the next release of FF.

**Comment by cadeyrn** [comment]
>  Since this is removed, the feature should show up in the next release of FF.
>
> The next release is scheduled for the 15th of November.

Assuming it will ride the trains it will be in Firefox 107 Beta and then Firefox 108, not in the stable release of Firefox 107 that will be released in November. Firefox 108 is on schedule for the 13th of December.

**Comment by SoftVision-LorandJanos** [comment]
There is no "Sort open tabs" section in the tab settings in the latest Nightly 108.0a1 (2022-11-03).
Will recheck later.

**Comment by Mugurell** [comment]
> There is no "Sort open tabs" section in the tab settings in the latest Nightly 108.0a1 (2022-11-03). Will recheck later.

The feature ended up having no setting.
There is no automatic functionality, users can do the "long press and drag tabs to reorder" gesture or not.
This was already enabled in Nightly, the recent patch enabled the feature in beta and release also starting with version 108.

**Comment by SoftVision-LorandJanos** [comment]
With this new information, I can confirm that the tabs can be manually rearranged by "long press and drag" in the latest Nightly 108.0a1 (2022-11-03).
Device used: Oppo Find X5 (Android 12).
Closing the ticket as verified.

**Comment by SoftVision-LorandJanos** [comment]
Verified as fixed on the latest Beta 108.0b1.
Device used: Google Pixel 4 (Android 13).

**Comment by MasterInQuestion** [comment]
Not possible for private tabs.

**Comment by Jorengarenar** [comment]
@MasterInQuestion I think you may need to create new issue for that

**Comment by MasterInQuestion** [comment]
Creating a new Issue for one essentially identical?..

**Comment by Jorengarenar** [comment]
> Creating a new Issue for one essentially identical?..

This one ain't gonna reopen itself and nobody gonna rebase the repo to change commit messages to undo "fixed by".  
So yes, this _feature request_ is complete. If you wanna have the _bug_ fixed, report it.
